rust/src/interpreter/tx_context.rs

```diff
@@ -0,0 +1,79 @@
+//! This module holds the [ExecutionTxContext].
+//! This type is the same as [evmc_vm::ExecutionTxContext] except that the pointer and length fields
+//! are converted to slices.
+//!
+//! Ideally this would be done already in [evmc_vm].
+use std::slice;
+
+use evmc_vm::{
+    ffi::{evmc_tx_context, evmc_tx_initcode},
+    Address, Uint256,
+};
+
+/// EVMC transaction context.
+#[derive(Debug, Copy, Clone, Hash, PartialEq)]
+pub struct ExecutionTxContext<'a> {
+    /// The transaction gas price.
+    pub tx_gas_price: Uint256,
+    /// The transaction origin account.
+    pub tx_origin: Address,
+    /// The miner of the block.
+    pub block_coinbase: Address,
+    /// The block number.
+    pub block_number: i64,
+    /// The block timestamp.
+    pub block_timestamp: i64,
+    /// The block gas limit.
+    pub block_gas_limit: i64,
+    /// The block previous RANDAO (EIP-4399).
+    pub block_prev_randao: Uint256,
+    /// The blockchain's ChainID.
+    pub chain_id: Uint256,
+    /// The block base fee per gas (EIP-1559, EIP-3198).
+    pub block_base_fee: Uint256,
+    /// The blob base fee (EIP-7516).
+    pub blob_base_fee: Uint256,
+    /// The array of blob hashes (EIP-4844).
+    pub blob_hashes: &'a [Uint256],
+    /// The array of transaction initcodes (TXCREATE).
+    pub initcodes: &'a [evmc_tx_initcode],
+}
+
+impl<'a> From<&'a evmc_vm::ExecutionTxContext> for ExecutionTxContext<'a> {
+    fn from(context: &'a evmc_tx_context) -> Self {
+        let blob_hashes = if context.blob_hashes.is_null() {
+            &[]
+        } else {
+            assert_ne!(context.blob_hashes_count, 0);
```

Comment:
We try to avoid assertions in the code for reasons we can discuss offline one day. In this case and the one below, I assume those could be easily avoided by just handling the empty case and the null case the same, right?

Reply:
Yes, that is true. However the expectation is that those slices are either empty (pointer is null and length is 0) or non empty (non zero pointer and length > 0). Having pointer = null and length > 0 violates this assumption and (at least in my mind) is a bug on the Go side. But I guess you prefer continuing with an empty slice over crashing. I'm fine with that too.

Reply:
I agree with @LorenzSchueler, receiving a null ptr from the FFI is unexpected and can be though of as a bug. Nevertheless, there is a critical responsibility on not panicking. I see two ways out:
- return an error right away. 
- continue in safety and let someone else figure out that there was an error.

Reply:
But if you use TryFrom what do you do in the calling rust code with this error? There is now proper way to continue but also no way to signal to Go that the expectations of the evmc interface are violated. In the ffi code which I adapted from the evmc rust bindings they just abort if something like this happens. Even on the go side there is (as far as I know) now way to deal with this because it is a bug in the go code itself.

Reply:
In my mind the only viable options are either to panic/ abort or to simply ignore this error completely and continue with an empty slice. The former would at least make sure that transactions are not processed incorrectly.

Reply:
Should I merge this as is or return an empty slice?

Reply:
Let's return the empty slice for now, just to prevent the inherent panic.